### PR TITLE
feat(conversation-manager): improve tool result truncation strategy

### LINKED ISSUE #1545: [FEATURE] Improve tool result truncation strategy for graceful context reduction
### Problem Statement

The current `SlidingWindowConversationManager` truncation logic has several issues:

1. **Truncates newest tool results first** - The `_find_last_message_with_tool_results` method searches from newest to oldest and truncates the most recent tool result. This is problematic because:
   - The newest tool result is likely the most relevant to the current task
   - Agent may interpret the truncated result (`"The tool result was too large!"` with `status: error`) as a tool failure
   - This can cause the agent to repeatedly call the same tool or make incorrect decisions based on misleading error status

2. **No size check before truncation** - Tool results are truncated regardless of their actual size. Small tool results may be unnecessarily truncated.

3. **Complete content removal** - The entire tool result content is replaced, losing all context about what the result contained.

### Proposed Solution

Implement a more graceful truncation strategy:

1. **Process older messages first** - Truncate oldest tool results before newest ones, as older results have likely already been processed and acted upon by the agent.

2. **Size-based threshold** - Only truncate content blocks that exceed a configurable character threshold.

3. **Partial truncation with context preservation** - Instead of removing entire content, preserve the beginning and end while truncating the middle:
   ```
   [first 200 chars of original content...]

   ... [truncated: 15000 chars removed] ...

   [...last 200 chars of original content]
   ```

4. **Image placeholder for older messages** - Replace images in older messages with descriptive placeholders:
   ```
   [image: filename.png, 1024x768]
   ```

### Use Case

1. **Prevent agent confusion** - Agent won't misinterpret truncated recent results as failures
2. **Maintain task continuity** - Recent tool results remain intact for ongoing work
3. **Preserve partial context** - Truncated content still provides hints about what information was available
4. **Reduce unnecessary retries** - Agent won't repeatedly call tools thinking previous calls failed

### Alternatives Solutions

**Separate into dedicated ConversationManager:**
- Extract truncation/compaction logic into a new `CompactingConversationManager`
- Keep `SlidingWindowConversationManager` focused solely on message count management (sliding window)
- Users can combine both managers as needed, following Single Responsibility Principle

### Additional Context

I've implemented and tested this approach in a separate project with positive results (67% token reduction, improved retention). 

Details: [Link](https://medium.com/towards-artificial-intelligence/long-context-compaction-for-ai-agents-part-2-implementation-and-evaluation-d708d2d2a6e5)

### PR DESCRIPTION
**Pull request by kevmyung**
## Description

Improves the `SlidingWindowConversationManager` truncation strategy to reduce
context size more gracefully without misleading the agent.

- Partially truncate large tool results (>400 chars) preserving first/last 200 chars,
  instead of replacing entire content with an error message
- Replace image blocks (top-level and inside toolResult) with descriptive text placeholders
- Target oldest tool results first so the most relevant recent results are preserved longer
- Add `ValueError` for invalid `per_turn` values (0 or negative integer)
- Rename `_find_last_message_with_tool_results` → `_find_oldest_message_with_tool_results`
  to accurately reflect iteration order

## Related Issues

Closes #1545

## Documentation PR

N/A

## Type of Change

Other: Enhancement - improve existing truncation behavior

## Testing

- [x] I ran `hatch run prepare`

Added unit tests covering partial truncation, image block replacement,
oldest-first ordering, idempotency guard, and boundary conditions.

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: src/strands/agent/conversation_manager/sliding_window_conversation_manager.py [hunk 1]
```diff
@@ -7,12 +7,15 @@
...
 from ...hooks import BeforeModelCallEvent, HookRegistry
-from ...types.content import Messages
+from ...types.content import ContentBlock, Messages
 from ...types.exceptions import ContextWindowOverflowException
+from ...types.tools import ToolResultContent
 from .conversation_manager import ConversationManager
 
 logger = logging.getLogger(__name__)
 
+_PRESERVE_CHARS = 200
```
Comment (JackYPCOnline) [inline src/strands/agent/conversation_manager/sliding_window_conversation_manager.py:17]: Question: any reason we don't want to make it configurable ?

Comment (kevmyung) [reply src/strands/agent/conversation_manager/sliding_window_conversation_manager.py:17]: Intentionally kept it as a constant to keep the API simple — any value here is a heuristic anyway. Happy to add it as a parameter if needed!

Comment (kevmyung) [reply src/strands/agent/conversation_manager/sliding_window_conversation_manager.py:17]: More specifically, exposing `preserve_chars` would also require exposing the truncation `threshold` for consistency. I thought both are heuristics that most users won't need to tune.

File: src/strands/agent/conversation_manager/sliding_window_conversation_manager.py [hunk 2]
```diff
@@ -214,50 +236,90 @@ def _truncate_tool_results(self, messages: Messages, msg_idx: int) -> bool:
...
+                tool_result_items = tool_result.get("content", [])
+                new_items: list[ToolResultContent] = []
+                item_changed = False
+
+                for item in tool_result_items:
+                    # Replace image items nested inside toolResult content
+                    if "image" in item:
+                        image = item["image"]
+                        media_type = image.get("format", "unknown")
+                        source = image.get("source", {})
```
Comment (JackYPCOnline) [inline src/strands/agent/conversation_manager/sliding_window_conversation_manager.py]: can we extract a helper function to reduce duplicate code?
